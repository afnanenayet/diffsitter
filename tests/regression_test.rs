#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use libdiffsitter::{
        diff::compute_edit_script, generate_ast_vector_data, input_processing::TreeSitterProcessor,
        parse::GrammarConfig,
    };
    use std::path::PathBuf;
    use test_case::test_case;

    /// Get paths to input files for tests
    fn get_test_paths(test_type: &str, test_name: &str, ext: &str) -> (PathBuf, PathBuf) {
        let test_data_root = PathBuf::from(format!("./test_data/{test_type}/{test_name}"));
        let path_a = test_data_root.join(format!("a.{ext}"));
        let path_b = test_data_root.join(format!("b.{ext}"));
        assert!(
            path_a.exists(),
            "test data path {} does not exist",
            path_a.to_str().unwrap()
        );
        assert!(
            path_b.exists(),
            "test data path {} does not exist",
            path_b.to_str().unwrap()
        );

        (path_a, path_b)
    }

    #[test_case("short", "rust", "rs", true, true)]
    #[test_case("short", "python", "py", true, true)]
    #[test_case("short", "go", "go", true, true)]
    #[test_case("medium", "rust", "rs", true, false)]
    #[test_case("medium", "rust", "rs", false, false)]
    #[test_case("medium", "cpp", "cpp", true, true)]
    #[test_case("medium", "cpp", "cpp", false, true)]
    fn diff_hunks_snapshot(
        test_type: &str,
        name: &str,
        ext: &str,
        split_graphemes: bool,
        strip_whitespace: bool,
    ) {
        let (path_a, path_b) = get_test_paths(test_type, name, ext);
        let config = GrammarConfig::default();
        let ast_data_a = generate_ast_vector_data(path_a, None, &config).unwrap();
        let ast_data_b = generate_ast_vector_data(path_b, None, &config).unwrap();

        let processor = TreeSitterProcessor {
            split_graphemes,
            strip_whitespace,
            ..Default::default()
        };

        let diff_vec_a = processor.process(&ast_data_a.tree, &ast_data_a.text);
        let diff_vec_b = processor.process(&ast_data_b.tree, &ast_data_b.text);
        let diff_hunks = compute_edit_script(&diff_vec_a, &diff_vec_b).unwrap();

        // We have to set the snapshot name manually, otherwise there appear to be threading issues
        // and we end up with more snapshot files than there are tests, which cause
        // nondeterministic errors.
        let snapshot_name = format!("{test_type}_{name}_split_graphemes_{split_graphemes}_strip_whitespace_{strip_whitespace}");
        assert_debug_snapshot!(snapshot_name, diff_hunks);
    }
}
