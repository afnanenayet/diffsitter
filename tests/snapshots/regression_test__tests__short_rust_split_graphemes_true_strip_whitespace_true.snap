---
source: tests/regression_test.rs
expression: snapshot_string
---
New(Line={line_index=9, entries=
[
    Entry{'}', start=(9, 0), end=(9, 1)}
]}
)
New(Line={line_index=11, entries=
[
    Entry{'f', start=(11, 0), end=(11, 1)}
    Entry{'n', start=(11, 1), end=(11, 2)}
    Entry{'a', start=(11, 3), end=(11, 4)}
    Entry{'d', start=(11, 4), end=(11, 5)}
    Entry{'d', start=(11, 5), end=(11, 6)}
    Entry{'i', start=(11, 6), end=(11, 7)}
    Entry{'t', start=(11, 7), end=(11, 8)}
    Entry{'i', start=(11, 8), end=(11, 9)}
    Entry{'o', start=(11, 9), end=(11, 10)}
    Entry{'n', start=(11, 10), end=(11, 11)}
    Entry{'(', start=(11, 11), end=(11, 12)}
    Entry{')', start=(11, 12), end=(11, 13)}
    Entry{'{', start=(11, 14), end=(11, 15)}
]}
)
Old(Line={line_index=1, entries=
[
    Entry{'l', start=(1, 4), end=(1, 5)}
    Entry{'e', start=(1, 5), end=(1, 6)}
    Entry{'t', start=(1, 6), end=(1, 7)}
    Entry{'x', start=(1, 8), end=(1, 9)}
    Entry{'=', start=(1, 10), end=(1, 11)}
    Entry{'1', start=(1, 12), end=(1, 13)}
    Entry{';', start=(1, 13), end=(1, 14)}
]}
)
New(Line={line_index=14, entries=
[
    Entry{'t', start=(14, 7), end=(14, 8)}
    Entry{'w', start=(14, 8), end=(14, 9)}
    Entry{'(', start=(14, 10), end=(14, 11)}
    Entry{')', start=(14, 11), end=(14, 12)}
]}
)
Old(Line={line_index=4, entries=
[
    Entry{'n', start=(4, 8), end=(4, 9)}
    Entry{'e', start=(4, 9), end=(4, 10)}
]}
)
